Improve-cross project actions - Edit via Grid

cnathe opened the pull request:
#### Rationale
As part of the effort to make "cross folder" action improvements in the apps, this set of PRs takes on the "Edit in Grid" action. This will allow for users with proper permission to edit selected rows in a sample, source, and assay runs/results grid when the set of selections includes rows from multiple containers.

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1472
- https://github.com/LabKey/labkey-ui-premium/pull/386
- https://github.com/LabKey/platform/pull/5427
- https://github.com/LabKey/limsModules/pull/141

#### Changes
- rename getOperationNotPermittedMessage() to getOperationNotAllowedMessage()
- EditableGridLoaderFromSelection to account for idsNotPermitted when using selections for getSelectedData()
- Show Project column in EditableGrid for "Edit in Grid"
- Add saveRowsByContainer util to QueryAPIWrapper
- EditableGrid column actions for add,remove,update should use editorModel.columns to determine column indices
- EditableGridPanelForUpdate props for readOnlyColumns and updateColumns, include originalRows as param to updateRows, show warning about notPermitted rows
- add getOperationConfirmationDataForModel helper
- fix in getUpdatedDataFromGrid for multi-value columns
- EditableGrid replace react-bootstrap OverlayTrigger/Popover with LabelHelpTip

cnathe committed:
rename getOperationNotPermittedMessage() to getOperationNotAllowedMessage()

cnathe committed:
EditableGridLoaderFromSelection to account for idsNotPermitted when using selections for getSelectedData()

cnathe committed:
EditableGridLoaderFromSelection to account for idsNotPermitted when using selections for getSelectedData()

cnathe committed:
3.37.2-fb-crossFolderEditInGrid.0

cnathe committed:
Show Project column in EditableGrid for "Edit in Grid"

cnathe committed:
3.37.2-fb-crossFolderEditInGrid.1

cnathe committed:
ChartBuilderModal fix for JS error on cancel

cnathe committed:
revert ChartBuilderModal.tsx JS fix (will target separate PR)

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.37.4-fb-crossFolderEditInGrid.0

cnathe committed:
Add saveRowsByContainer util to QueryAPIWrapper

cnathe committed:
3.37.4-fb-crossFolderEditInGrid.2

cnathe committed:
EditableGrid column actions for add,remove,update should use editorModel.columns to determine column indices

cnathe committed:
3.37.4-fb-crossFolderEditInGrid.3

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.0

cnathe committed:
add const for ASSAY_RESULTS_OPERATION_CONFIRMATION_ACTION

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.1

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.2

cnathe committed:
EditableGridPanelForUpdate props for readOnlyColumns and updateColumns

cnathe committed:
EditableGridPanelForUpdate include originalRows as param to updateRows

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.3

cnathe committed:
missed from prev commit

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.4

cnathe committed:
EditableGridPanelForUpdate to show warning about notPermitted rows

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.5

cnathe committed:
add DataOperation.Edit

cnathe committed:
3.37.5-fb-crossFolderEditInGrid.6

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.37.7-fb-crossFolderEditInGrid.0

cnathe committed:
add getOperationConfirmationDataForModel helper

cnathe committed:
3.37.7-fb-crossFolderEditInGrid.1

cnathe committed:
fix in getUpdatedDataFromGrid for multi-value columns

cnathe committed:
3.37.7-fb-crossFolderEditInGrid.2

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.38.0-fb-crossFolderEditInGrid.0

cnathe committed:
EditableGrid replace react-bootstrap OverlayTrigger/Popover with LabelHelpTip

cnathe committed:
update release notes

labkey-nicka commented on packages/components/src/internal/components/editable/EditableGridPanelForUpdate.tsx:
```diff
@@ -60,6 +77,7 @@ export const EditableGridPanelForUpdate: FC<Props> = props => {
     const { requiresUserComment } = useDataChangeCommentsRequired();
 
     const hasValidUserComment = comment?.trim()?.length > 0;
+    const notPermittedText = getOperationNotPermittedMessage(editStatusData, pluralNoun);
```
nit: Consider memoizing.

```ts
const notPermittedText = useMemo<string>(
    () => getOperationNotPermittedMessage(editStatusData, pluralNoun),
    [editStatusData, pluralNoun]
);
```

cnathe replied:
done

labkey-nicka commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -581,7 +575,7 @@ export function removeColumn(
     fieldKey: string
 ): EditorModelUpdates {
-    const deleteIndex = queryInfo.getInsertColumnIndex(fieldKey);
+    const deleteIndex = editorModel.columns.findIndex(column => column === fieldKey);
```
The previous method on `QueryInfo` attempted to be case-insensitive by doing a `toLowerCase()` comparison check. In general, I think we'd be better off having the `columns: List<string>` be guaranteed to be lowercase so we don't run into weird casing conflicts.

cnathe replied:
I saw that some of the methods in that file were using `Utils.caseInsensitiveEquals()`. I updated all of the `column.findIndex` cases to use that as well.

labkey-nicka commented on packages/components/src/internal/components/editable/actions.ts:
```diff
@@ -1301,7 +1295,7 @@ function parsePastedLookup(column: QueryColumn, descriptors: ValueDescriptor[],
             if (vt.length > 0) {
                 const vl = vt.toLowerCase();
-                const vd = descriptors.find(d => d.display && d.display.toString().toLowerCase() === vl);
+                const vd = descriptors?.find(d => d.display && d.display.toString().toLowerCase() === vl);
```
This `?` is a code smell to be. This method should not be receiving `undefined`/`null` descriptors.

cnathe replied:
Good point. The repro here was trying to paste a value into the editable grid for the read-only Project/Folder column, which is a lookup. So this code was trying to `parsePastedLookup` for the lookup column even though it will never use the value. I instead made a change in `insertPastedData` so that we check `!readOnlyCol && !readOnlyCell` first.

labkey-nicka commented on packages/components/src/internal/query/api.ts:
```diff
@@ -994,6 +994,39 @@ export function saveRows(options: Query.SaveRowsOptions): Promise<Query.SaveRows
+    return containerRows;
+}
+
+export function saveRowsByContainer(options: Query.SaveRowsOptions, containerField: string = 'ContainerPath'): void {
```
A couple things:

1. This should make use of `saveRows` already declared rather than calling `Query.saveRows` directly.
2. Update the interface for the options to `saveRows` to exclude `failure` and `success`.
3. Update the return type of `saveRowsByContainer` to be `Promise<Query.SaveRowsResponse>` so it can be handled async.

With this it'll be promise based

```ts
export type SaveRowsOptions = Omit<Query.SaveRowsOptions, 'failure' | 'success'>;

export function saveRows(options: SaveRowsOptions): Promise<Query.SaveRowsResponse> {
    return new Promise((resolve, reject) => {
        Query.saveRows({
            apiVersion: 13.2,
            ...options,
            success: response => {
                resolve(response);
            },
            failure: error => {
                reject(error);
            },
        });
    });
}

export function saveRowsByContainer(
    options: SaveRowsOptions,
    containerField: string = 'ContainerPath'
): Promise<Query.SaveRowsResponse> {
    const commands: Query.Command[] = [];

    // for each original command, split it into multiple commands for each container in the rows
    options.commands.forEach(command => {
        const containerRows = splitRowsByContainer(command.rows, containerField);
        Object.keys(containerRows).forEach(containerPath => {
            const rows = containerRows[containerPath];
            commands.push({
                ...command,
                rows,
                containerPath: !containerPath || containerPath === 'undefined' ? undefined : containerPath,
            });
        });
    });

    return saveRows({ ...options, commands });
}
```

cnathe replied:
Good suggestion. done.

labkey-nicka commented on packages/components/src/internal/query/api.ts:
```diff
@@ -994,6 +994,39 @@ export function saveRows(options: Query.SaveRowsOptions): Promise<Query.SaveRows
+}
+
+export function saveRowsByContainer(options: Query.SaveRowsOptions, containerField: string = 'ContainerPath'): void {
+    const commands = [];
```
This should be typed as 

```ts
const commands: Query.Command[] = [];
```

Doing to exposes the fact that `containerPath` is not configured as part of the interface in `@labkey/api`.

labkey-nicka commented on packages/components/src/internal/query/api.ts:
```diff
@@ -994,6 +994,39 @@ export function saveRows(options: Query.SaveRowsOptions): Promise<Query.SaveRows
+            commands.push({
+                ...command,
+                rows,
+                containerPath: !containerPath || containerPath === 'undefined' ? undefined : containerPath,
```
Seems odd we need to check here for the string `'undefined``. Why is this? Likely, worth a comment.

cnathe replied:
comment added

labkey-nicka commented on packages/components/src/internal/query/APIWrapper.ts:
```diff
@@ -98,6 +99,7 @@ export interface QueryAPIWrapper {
         inherit: boolean,
         shared: boolean
     ) => Promise<void>;
+    saveRowsByContainer: (options: Query.SaveRowsOptions, containerField: string) => void;
```
I try to avoid passing along the notion of `success`/`failure` handlers by only doing that in the wrapped method and then excluding them from the options interface. This notion leaks everywhere if we don't.

labkey-nicka replied:
Duplicate of https://github.com/LabKey/labkey-ui-components/pull/1472#discussion_r1561271902. Resolving.

labkey-nicka commented on packages/components/src/internal/query/api.ts:
```diff
@@ -994,6 +994,39 @@ export function saveRows(options: Query.SaveRowsOptions): Promise<Query.SaveRows
+    const commands = [];
+
+    // for each original command, split it into multiple commands for each container in the rows
+    options.commands.forEach(command => {
```
nit: Seems a bit odd that the `containerField` defaults to `"ContainerPath"` which is not a normal field in LKS. I would expect maybe `Folder/Path`?

cnathe replied:
changed default to `'Folder'`

cnathe committed:
CR feedback:
- use Utils.caseInsensitiveEquals() in add/remove/update columns
- saveRowsByContainer fixes for typings and handling of promise

cnathe committed:
3.38.0-fb-crossFolderEditInGrid.1

labkey-nicka commented on packages/components/src/internal/components/editable/EditableGridPanelForUpdate.tsx:
```diff
@@ -31,21 +40,29 @@ type Models = {
 
 interface Props {
     containerFilter?: Query.ContainerFilter;
+    editStatusData?: OperationConfirmationData;
```
nit: Consider altering this interface to actually extend `EditGridPanelProps`:

```tsx
type InheritedEditableGridPanelProps = Omit<
    EditableGridPanelProps,
    | 'columnMetadata'
    | 'editorModel'
    | 'forUpdate'
    | 'model'
    | 'onChange'
>;

interface EditableGridPanelForUpdateProps extends InheritedEditableGridPanelProps {
    editStatusData?: OperationConfirmationData;
    idField: string;
    loader: EditableGridLoader;
    onCancel: () => void;
    onComplete: () => void;
    pluralNoun?: string;
    queryModel: QueryModel;
    selectionData: Map<string, any>;
    singularNoun?: string;
    updateRows: (
        schemaQuery: SchemaQuery,
        rows: Array<Record<string, any>>,
        comment: string,
        originalRows: Array<Record<string, any>>
    ) => Promise<any>;
}
```

then the prop loadout within the component becomes:

```tsx
const { editStatusData, idField, loader, onCancel, onComplete, pluralNoun, queryModel, selectionData, singularNoun, updateRows, ...editableGridProps } = props;
const { getIsDirty, setIsDirty } = editableGridProps;
```

and finally the component is initialized with the proper overrides:

```tsx
<EditableGridPanel
    allowAdd={false}
    allowRemove={false}
    bordered
    bsStyle="info"
    striped
    title={`Edit selected ${pluralNoun}`}
    {...editableGridProps}
    columnMetadata={columnMetadata}
    editorModel={models.editorModel}
    forUpdate
    model={models.dataModel}
    onChange={onGridChange}
/>
```

cnathe replied:
done

cnathe committed:
add getOperationConfirmationDataForModel to EntityAPIWrapper

cnathe committed:
inherit EditableGridPanelForUpdateProps from EditableGridPanelProps

cnathe committed:
add getCrossFolderSelectionResult to EntityAPIWrapper

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.38.1-fb-crossFolderEditInGrid.0

cnathe committed:
Editable grid error to scroll into view when shown below grid

cnathe committed:
no longer need hiddenFieldNames for PlateTemplate

cnathe committed:
3.38.1-fb-crossFolderEditInGrid.1
3.38.1-fb-crossFolderEditInGrid.1

cnathe committed:
jest snapshot update

cnathe committed:
Issue 50140: use defaultValue for alias name input to prevent "cursor jump to end" on change

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
3.38.2-fb-crossFolderEditInGrid.0

cnathe committed:
jest fixes related to issue 50140

cnathe committed:
remove DatasetDesignerPanels snapshot test

cnathe committed:
Fix for preventing more editable grid actions on readOnly cells

cnathe committed:
3.38.2-fb-crossFolderEditInGrid.1

cnathe committed:
QueryInfo not to warn on missing default view in test

cnathe committed:
3.38.2-fb-crossFolderEditInGrid.2

cnathe committed:
revert

cnathe committed:
remove console.log

cnathe committed:
fix to allow adding column at left if fieldKey doesn't exist in addColumns()

cnathe committed:
3.38.2-fb-crossFolderEditInGrid.3

cnathe committed:
new jest tests for misc utils

cnathe committed:
jest test for splitRowsByContainer

cnathe committed:
3.38.2-fb-crossFolderEditInGrid.4

cnathe committed:
Merge branch 'develop' into fb_crossFolderEditInGrid

# Conflicts:
#	packages/components/package-lock.json
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
Update release notes with version number and release date

cnathe committed:
3.39.0

cnathe committed:
npm run lint-branch-fix
